Description: [PATCH 0001/0006] Add aa profile post-process support
 Set the stage for distro specific apparmor profiles.
 This is needed sicne the apparmor version has different features in
 Debian/Ubuntu. Yet we want to add some version dependent rules.
 To do so the apparmor examples get preprocessed according to a target.
Author: Stefan Bader <stefan.bader@canonical.com>
Forwarded: no

--- a/configure.ac
+++ b/configure.ac
@@ -1365,6 +1365,28 @@
 fi
 AM_CONDITIONAL([WITH_APPARMOR_PROFILES], [test "$with_apparmor_profiles" != "no"])
 
+AC_ARG_WITH([apparmor-profiles-version],
+  [AS_HELP_STRING([--with-apparmor-profiles-version],
+    [install apparmor profiles for apparmor version @<:@default=check@:>@])],
+  [],
+  [with_apparmor_profiles_version=check])
+if test "$with_apparmor_profiles" = "no"; then
+  with_apparmor_profiles_version="no"
+fi
+if test "$with_apparmor_profiles_version" = "check"; then
+  APPARMOR_VERSION=`pkg-config --modversion libapparmor|cut -d. -f1-2`
+elif test "$with_apparmor_profiles_version" != "no"; then
+  APPARMOR_VERSION=$withval
+fi
+if test "$with_apparmor_profiles_version" != "no"; then
+  APPARMOR_MAJOR_VERSION=`echo $APPARMOR_VERSION|cut -d. -f1`
+  APPARMOR_MINOR_VERSION=`echo $APPARMOR_VERSION|cut -d. -f2`
+  APPARMOR_VERSION_NUMBER=`expr $APPARMOR_MAJOR_VERSION \* 1000 + $APPARMOR_MINOR_VERSION`
+  AC_DEFINE_UNQUOTED([APPARMOR_VERSION_NUMBER],
+    $APPARMOR_VERSION_NUMBER,
+    [Version number of apparmor library (for profile features)])
+fi
+
 dnl DTrace static probes
 AC_ARG_WITH([dtrace],
   [AS_HELP_STRING([--with-dtrace],
--- a/examples/Makefile.am
+++ b/examples/Makefile.am
@@ -18,13 +18,12 @@
 
 FILTERS = $(wildcard $(srcdir)/xml/nwfilter/*.xml)
 
+CLEANFILES =
+BUILT_SOURCES =
+
 EXTRA_DIST = \
 	apparmor/TEMPLATE.qemu \
 	apparmor/TEMPLATE.lxc \
-	apparmor/libvirt-qemu \
-	apparmor/libvirt-lxc \
-	apparmor/usr.lib.libvirt.virt-aa-helper \
-	apparmor/usr.sbin.libvirtd \
 	lxcconvert/virt-lxc-convert \
 	polkit/libvirt-acl.rules \
 	$(wildcard $(srcdir)/systemtap/*.stp) \
@@ -67,6 +66,12 @@
 admin_client_close_SOURCES = admin/client_close.c
 
 if WITH_APPARMOR_PROFILES
+BUILT_SOURCES += \
+	apparmor/usr.lib.libvirt.virt-aa-helper \
+	apparmor/usr.sbin.libvirtd \
+	apparmor/libvirt-qemu \
+	apparmor/libvirt-lxc
+
 apparmordir = $(sysconfdir)/apparmor.d/
 apparmor_DATA = \
 	apparmor/usr.lib.libvirt.virt-aa-helper \
@@ -84,6 +89,36 @@
 	apparmor/TEMPLATE.qemu \
 	apparmor/TEMPLATE.lxc \
 	$(NULL)
+
+apparmor/stamp:
+	test -d apparmor || mkdir apparmor
+	touch apparmor/stamp
+
+apparmor/libvirt-lxc:   $(srcdir)/apparmor/libvirt-lxc.in \
+			$(srcdir)/apparmor/profile-preprocess \
+			../config.h apparmor/stamp
+	sh $(srcdir)/apparmor/profile-preprocess $< >$@
+
+apparmor/libvirt-qemu:	$(srcdir)/apparmor/libvirt-qemu.in \
+			$(srcdir)/apparmor/profile-preprocess \
+			../config.h apparmor/stamp
+	sh $(srcdir)/apparmor/profile-preprocess $< >$@
+
+apparmor/usr.lib.libvirt.virt-aa-helper: \
+			$(srcdir)/apparmor/usr.lib.libvirt.virt-aa-helper.in \
+			$(srcdir)/apparmor/profile-preprocess \
+			../config.h apparmor/stamp
+	sh $(srcdir)/apparmor/profile-preprocess $< >$@
+
+apparmor/usr.sbin.libvirtd: \
+			$(srcdir)/apparmor/usr.sbin.libvirtd.in \
+			$(srcdir)/apparmor/profile-preprocess \
+			../config.h apparmor/stamp
+	sh $(srcdir)/apparmor/profile-preprocess $< >$@
+
+CLEANFILES += apparmor/libvirt-lxc apparmor/libvirt-qemu
+CLEANFILES += apparmor/usr.lib.libvirt.virt-aa-helper
+CLEANFILES += apparmor/usr.sbin.libvirtd
 endif WITH_APPARMOR_PROFILES
 
 if WITH_NWFILTER
--- /dev/null
+++ b/examples/apparmor/profile-preprocess
@@ -0,0 +1,26 @@
+#!/bin/sh
+
+PROFILES_VERSION=$(
+	awk '$1=="#define" && $2=="APPARMOR_VERSION_NUMBER"{
+		print $3
+	}' ../config.h)
+
+awk -vVERSION=$PROFILES_VERSION '
+$1 == "@@ifge" {
+	if (VERSION < $2)
+		skip=1
+	next
+}
+$1 == "@else" && skip == 1{
+	skip=0
+	next
+}
+$1 == "@@end"{
+	skip=0
+	next
+}
+!skip{
+	print
+}' $1
+
+exit 0
